File: app/src/main/java/com/ddiehl/android/flippit/game/Board.java
package com.ddiehl.android.flippit.game;


import android.content.Context;

import com.ddiehl.android.flippit.utils.BoardIterator;

public class Board {
    private static final String TAG = Board.class.getSimpleName();
	private static Board _instance = null;
    private static Context context;
	private BoardSpace[][] spaces;
	private int width;
	private int height;

    private final int[][] moveDirections = new int[][] {
            {0,  -1}, // Down
            {1,  0}, // Right
            {-1, 0}, // Left
            {0,  1}, // Up
            {-1, -1}, // Down-Left
            {1,  -1}, // Down-Right
            {-1, 1}, // Top-Left
            {1,  1} // Top-Right
    };

	private Board(Context c) {
        context = c;
		width = 8;
		height = 8;
        spaces = new BoardSpace[height][width];
		reset();
	}

    protected Board copy() {
        Board copy = new Board(context);
        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                copy.spaces[y][x] = spaces[y][x].copy(context);
            }
        }
        return copy;
    }

	public void reset() {
		for (int y = 0; y < height; y++) {
			for (int x = 0; x < width; x++) {
				spaces[y][x] = new BoardSpace(context, x, y);
			}
		}

		spaces[3][3].setColor(ReversiColor.White);
		spaces[3][4].setColor(ReversiColor.Black);
		spaces[4][4].setColor(ReversiColor.White);
		spaces[4][3].setColor(ReversiColor.Black);
	}

	public boolean hasMove(Player p) {
		BoardIterator i = new BoardIterator(this);
		while (i.hasNext()) {
			BoardSpace s = i.next();
			if (s.isOwned())
				continue;
			for (int[] move : moveDirections) {
				int value = moveValueInDirection(s, move[0], move[1], p.getColor());
				if (value != 0) {
					return true;
				}
			}
		}
		return false;
	}

	public BoardSpace getSpaceAt(int x, int y) {
		if (x >= 0 && x < width && y >= 0 && y < height)
			return spaces[y][x];

		return null;
	}

	public boolean commitPiece(BoardSpace space, ReversiColor playerColor) {
		for (int[] move : moveDirections) {
			if (moveValueInDirection(space, move[0], move[1], playerColor) != 0) {
				flipInDirection(space, move[0], move[1], playerColor);
			}
		}
		return true;
	}

	public int spacesCapturedWithMove(BoardSpace s, ReversiColor playerColor) {
		int moveVal = 0;
		for (int[] move : moveDirections)
			moveVal += moveValueInDirection(s, move[0], move[1], playerColor);
		return moveVal;
	}

	public int moveValueInDirection(BoardSpace s, int dx, int dy, ReversiColor playerColor) {
		if (s.x+dx < 0 || s.x+dx >= width || s.y+dy < 0 || s.y+dy >= height)
			return 0;

		int moveVal = 0;
		ReversiColor opponentColor = (playerColor == ReversiColor.Black) ? ReversiColor.White : ReversiColor.Black;
        BoardSpace firstPiece = getSpaceAt(s.x + dx, s.y + dy);

		if (firstPiece != null && firstPiece.getColor() == opponentColor) {
			int cx = s.x+dx;
			int cy = s.y+dy;
			while (getSpaceAt(cx, cy) != null && getSpaceAt(cx, cy).getColor() == opponentColor) {
				moveVal++;
				cx += dx;
				cy += dy;
			}
			if (getSpaceAt(cx, cy) != null && getSpaceAt(cx, cy).getColor() == playerColor) {
				return moveVal;
			}
		}

		return 0;
	}

    public void flipInDirection(BoardSpace s, int dx, int dy, ReversiColor playerColor) {
        s.setColor(playerColor);
        int cx = s.x + dx;
        int cy = s.y + dy;

        while (getSpaceAt(cx, cy).getColor() != playerColor) {
            getSpaceAt(cx, cy).flipColor();
            cx += dx;
            cy += dy;
        }
    }

	public int width() {
		return width;
	}

	public int height() {
		return height;
	}

	public static Board getInstance(Context c) {
		if (_instance == null)
			_instance = new Board(c);

		return _instance;
	}

    public Board deserialize(Context context, String in) {
        int index = 0;
        for (int y = 0; y < _instance.height; y++) {
            for (int x = 0; x < _instance.width; x++) {
                char c = in.charAt(index);
                spaces[y][x] = new BoardSpace(context, x, y);
                switch (c) {
                    case '0':
                        break;
                    case '1':
                        spaces[y][x].setColor(ReversiColor.White);
                        break;
                    case '2':
                        spaces[y][x].setColor(ReversiColor.Black);
                }
            }
        }
        return this;
    }

    public String serialize() {
        StringBuilder sb = new StringBuilder();
		BoardIterator i = new BoardIterator(this);
		while (i.hasNext()) {
			BoardSpace s = i.next();
			if (!s.isOwned())
				sb.append("0");
			else if (s.getColor() == ReversiColor.White)
				sb.append("1");
			else
				sb.append("2");
		}
        return sb.toString();
    }

	@Override
	public String toString() {
		StringBuilder sb = new StringBuilder();
        sb.append("\n");
		for (int y = 0; y < height; y++) {
			for (int x = 0; x < width; x++) {
                BoardSpace s = getSpaceAt(x, y);
                if (!s.isOwned())
                    sb.append("0 ");
				else if (s.getColor() == ReversiColor.White)
					sb.append("1 ");
				else
					sb.append("2 ");
			}
			if (y != spaces.length-1)
				sb.append("\n");
		}
		return sb.toString();
	}
}

File: app/src/main/java/com/ddiehl/android/reversi/model/Board.kt
package com.ddiehl.android.reversi.model


import com.ddiehl.android.reversi.IllegalMoveException
import com.ddiehl.android.reversi.byteArrayToString
import rx.Observable
import rx.functions.Func0

class Board(val height: Int, val width: Int) : Iterable<BoardSpace> {

    data class MoveDirection(val dx: Int, val dy: Int)

    companion object {
        private val MOVE_DIRECTIONS = listOf(
                MoveDirection(0, -1), // Down
                MoveDirection(1, 0), // Right
                MoveDirection(-1, 0), // Left
                MoveDirection(0, 1), // Up
                MoveDirection(-1, -1), // Down-Left
                MoveDirection(1, -1), // Down-Right
                MoveDirection(-1, 1), // Top-Left
                MoveDirection(1, 1) // Top-Right
        )
    }

    fun restoreState(saved: ByteArray) {
        restoreState(byteArrayToString(saved))
    }

    fun restoreState(saved: String): Board {
        var index = 0
        for (y in 0 until height) {
            for (x in 0 until width) {
                val c = saved[index++]

                when (c) {
                    '0' -> { getSpaceAt(x, y).color = null }
                    '1' -> getSpaceAt(x, y).color = ReversiColor.LIGHT
                    '2' -> getSpaceAt(x, y).color = ReversiColor.DARK
                }
            }
        }
        return this
    }

    private val spaces: Array<Array<BoardSpace>> =
            Array(height) { row ->
                Array(width) { col ->
                    BoardSpace(row, col)
                }
            }

    fun reset() {
        // Set the color of each space to null first
        forEach { space -> space.color = null }

        // Then set the center 4 spaces to the starting configuration
        spaces[3][3].color = ReversiColor.LIGHT
        spaces[3][4].color = ReversiColor.DARK
        spaces[4][4].color = ReversiColor.LIGHT
        spaces[4][3].color = ReversiColor.DARK
    }

    fun copy(): Board {
        val copy = Board(height, width)
        for (y in 0 until height) {
            for (x in 0 until width) {
                copy.spaces[y][x] = spaces[y][x].copy()
            }
        }
        return copy
    }

    fun hasMove(color: ReversiColor): Boolean {
        forEach { space ->
            MOVE_DIRECTIONS
                    .filter { !space.isOwned }
                    .filter { (dx, dy) -> isWithinBounds(space.x + dx, space.y + dy) }
                    .map { (dx, dy) -> moveValueInDirection(space, dx, dy, color) }
                    .filter { value -> value > 0 }
                    .forEach { return true }
        }
        return false
    }

    fun isWithinBounds(x: Int, y: Int): Boolean {
        return x in 0..(width - 1) && y in 0..(height - 1)
    }

    fun getSpaceAt(x: Int, y: Int): BoardSpace =
            if (isWithinBounds(x, y)) {
                spaces[y][x]
            } else {
                throw NullPointerException("Space @ $x $y is null")
            }

    fun requestClaimSpace(row: Int, col: Int, color: ReversiColor): Observable<Boolean> {
        return Observable.defer(Func0 {
            val space = spaces[row][col]

            // If space is already claimed, return an error
            if (space.isOwned) {
                return@Func0 Observable.error<Boolean>(IllegalMoveException("space is already owned"))
            }

            val captured = spacesCapturedWithMove(space, color)
            if (captured <= 0) {
                return@Func0 Observable.error<Boolean>(IllegalMoveException("move value is <= 0: " + captured))
            }

            commitPiece(space, color)
            Observable.just(true)
        })
    }

    fun commitPiece(space: BoardSpace, playerColor: ReversiColor) {
        MOVE_DIRECTIONS
                .filter { (dx, dy) -> moveValueInDirection(space, dx, dy, playerColor) != 0 }
                .forEach { (dx, dy) -> flipInDirection(space, dx, dy, playerColor) }
    }

    internal fun spacesCapturedWithMove(space: BoardSpace, playerColor: ReversiColor): Int =
            MOVE_DIRECTIONS
                    .filter { space.color == null }
                    .filter { (dx, dy) -> isWithinBounds(space.x + dx, space.y + dy) }
                    .sumBy { (dx, dy) -> moveValueInDirection(space, dx, dy, playerColor) }

    private fun moveValueInDirection(space: BoardSpace, dx: Int, dy: Int, playerColor: ReversiColor): Int {
        // If the move would bring us out of bounds of the board area, just return 0
        if (!isWithinBounds(space.x + dx, space.y + dy)) {
            return 0
        }

        // Otherwise, calculate how many spaces we can capture in that direction
        var moveVal = 0
        val opponentColor = if (playerColor == ReversiColor.DARK) ReversiColor.LIGHT else ReversiColor.DARK
        val firstPiece = getSpaceAt(space.x + dx, space.y + dy)

        if (firstPiece.color == opponentColor) {
            var currentX = space.x + dx
            var currentY = space.y + dy

            while (isWithinBounds(currentX, currentY) && getSpaceAt(currentX, currentY).color == opponentColor) {
                moveVal++
                currentX += dx
                currentY += dy
            }

            if (isWithinBounds(currentX, currentY) && getSpaceAt(currentX, currentY).color == playerColor) {
                return moveVal
            }
        }

        return 0
    }

    private fun flipInDirection(space: BoardSpace, dx: Int, dy: Int, playerColor: ReversiColor) {
        space.color = playerColor
        var cx = space.x + dx
        var cy = space.y + dy

        while (getSpaceAt(cx, cy).color != playerColor) {
            getSpaceAt(cx, cy).flip()
            cx += dx
            cy += dy
        }
    }

    fun getNumSpacesForColor(color: ReversiColor) = count { it.isOwned && it.color == color }

    val numberOfEmptySpaces: Int
        get() = count { !it.isOwned }

    fun serialize(): ByteArray {
        val out = ByteArray(64)
        forEachIndexed { index, space ->
            when {
                !space.isOwned -> out[index] = 0
                space.color == ReversiColor.LIGHT -> out[index] = 1
                else -> out[index] = 2
            }
        }
        return out
    }

    fun getSpaceNumber(s: BoardSpace): Byte {
        return (s.y * 8 + s.x + 1).toByte()
    }

    fun getBoardSpaceFromNum(n: Int): BoardSpace {
        return getSpaceAt((n-1) % 8, (n-1) / 8)
    }

    fun spaceAt(row: Int, col: Int): BoardSpace {
        return spaces[row][col]
    }

    override fun iterator(): Iterator<BoardSpace> = BoardIterator(this)

    override fun toString(): String {
        val string = StringBuilder()
        string.append("\n")

        for (row in 0 until height) {
            (0 until width)
                    .map { col -> getSpaceAt(col, row) }
                    .forEach {
                        val char = when {
                            !it.isOwned ->
                                "0 "
                            it.color == ReversiColor.LIGHT ->
                                "1 "
                            else ->
                                "2 "
                        }
                        string.append(char)
                    }
            string.append(if (row != spaces.size - 1) "\n" else "")
        }

        return string.toString()
    }
}
